### ISSUE TITLE
[`filename-case`] Allow ignoring a directory segment without also skipping the filename check

### ISSUE THREAD

**Issue by TitikshaBansal** [sindresorhus/eslint-plugin-unicorn#3581]
## Real-world use case

This issue came out of a real GitLab project. Full context and the exact reproduction discussion here: [gitlab-org/.../contributors-gitlab-com#556](https://gitlab.com/gitlab-org/developer-relations/contributor-success/contributors-gitlab-com/-/work_items/556).

Short version: the project follows the Rails/Vite convention of putting Vue components under `app/javascript/…`. Since `eslint-plugin-unicorn` v65 the `filename-case` rule started validating directory segments too, so the lowercase `app` folder began failing PascalCase enforcement. The project shipped `checkDirectories: false` as a workaround ([MR !2348](https://gitlab.com/gitlab-org/developer-relations/contributor-success/contributors-gitlab-com/-/merge_requests/2348)), but that also disables directory casing checks everywhere else — the maintainer flagged this as unideal and asked whether a directory-only skip is possible. Investigation confirmed the current API can't express it, hence this issue.

## Problem

The `ignore` option in `unicorn/filename-case` is a full early-return: if any path segment matches an `ignore` regex, both the directory-name check *and* the filename check are skipped for that file.

This makes it impossible to express the following (fairly common) use case: **"validate PascalCase on all `.vue` files under `app/javascript/`, but don't complain that the `app` directory itself is lowercase."**

The `app/` folder name is a Rails / Vite convention and can't be renamed. Since v65, `filename-case` also validates directory names, so the only way to silence the false positive today is `checkDirectories: false` — which disables directory casing enforcement *everywhere*, not just for the `app` segment.

## Reproduction

`eslint.config.js`:

```js
{
  files: ['app/javascript/**/*.vue'],
  rules: {
    'unicorn/filename-case': ['error', {
      cases: { pascalCase: true },
      checkDirectories: true,
      ignore: [/^app$/],
    }],
  },
}
```

Create `app/javascript/pages/badly-named.vue`. Expected: rule flags the kebab-case filename. Actual: rule passes silently, because `pathSegments.some(segment => ignore.some(re => re.test(segment)))` matches on the `app` segment and returns before the filename is validated.

Relevant source (`rules/filename-case.js`):

```js
if (pathSegments.some(segment => ignore.some(regexp => regexp.test(segment)))) {
    return;  // skips directory AND filename checks
}
if (isCheckDirectories) { /* ... */ }
```

## Proposal

Introduce a way to exempt directory segments from casing enforcement *without* also skipping the filename check on files inside them. Two possible shapes:

**Option A — new `ignoreDirectories` option.**

```js
'unicorn/filename-case': ['error', {
  checkDirectories: true,
  ignoreDirectories: [/^app$/],  // only skips the directory segment check
  ignore: [/* still a full-rule bypass, unchanged */],
}]
```

Backwards compatible; `ignore` semantics preserved.

**Option B — change `ignore` to skip only the directory check when `checkDirectories: true`.**

Cleaner API but a breaking change. Users relying on `ignore` as a whole-file bypass (per #1406's original intent) would need to migrate.

Option A seems safer.

## Related

- #1406 — original `ignore` feature: intent was whole-file exemption based on path, so today's behavior is consistent with that intent.
- #686 — introduced `checkDirectories`.
- #2642 — related but different: asked for a blanket "off" inside `__mocks__`; this issue asks for a *directory-only* skip that still validates filenames.

Happy to send a PR if the maintainers are open to Option A.

**Comment by ltickett** [comment]
Thank you for creating the issue @TitikshaBansal 

Sorry if I missed it, but I didn’t see a reference to our project. I think it’s worth including it here to see the exact example/recreation steps you details above is a real world scenario https://gitlab.com/gitlab-org/developer-relations/contributor-success/contributors-gitlab-com/-/work_items/556

**Comment by TitikshaBansal** [comment]
> Thank you for creating the issue [@TitikshaBansal](https://github.com/TitikshaBansal)
> 
> Sorry if I missed it, but I didn’t see a reference to our project. I think it’s worth including it here to see the exact example/recreation steps you details above is a real world scenario https://gitlab.com/gitlab-org/developer-relations/contributor-success/contributors-gitlab-com/-/work_items/556

Great point, thanks @ltickett  — I've added a "Real-world use case" section to the issue body linking [contributors-gitlab-com#556](https://gitlab.com/gitlab-org/developer-relations/contributor-success/contributors-gitlab-com/-/work_items/556) and the related MR !2348 so the reproduction context is right at the top.

**Comment by sindresorhus** [comment]
Thanks for raising this. For the linked project, an ESLint file override already handles this:

```js
{
	files: ['app/javascript/**/*.vue'],
	rules: {
		'unicorn/filename-case': ['error', {
			cases: {pascalCase: true},
			checkDirectories: false,
		}],
	},
}
```

This preserves filename checking while allowing the existing directory structure. An `ignoreDirectories` option would only skip `app` and still report other directory names in the subtree. Could you share a case where this override is insufficient?

**Comment by TitikshaBansal** [comment]
> Thanks for raising this. For the linked project, an ESLint file override already handles this:
> 
> {
> 	files: ['app/javascript/**/*.vue'],
> 	rules: {
> 		'unicorn/filename-case': ['error', {
> 			cases: {pascalCase: true},
> 			checkDirectories: false,
> 		}],
> 	},
> }
> This preserves filename checking while allowing the existing directory structure. An `ignoreDirectories` option would only skip `app` and still report other directory names in the subtree. Could you share a case where this override is insufficient?

Thanks for the close look. You're right — I re-checked the linked project's tree and every subdirectory under `app/javascript/` (`common`, `pages`, `entrypoints`, `graphql`, `spec`, `stylesheets`) is also lowercase. With `pascalCase: true`, `ignoreDirectories: [/^app$/]` would flag all of them, so it doesn't actually improve on the scoped `checkDirectories: false` override for this project.

The original ask was made on the assumption that only `app` was non-compliant, which turned out not to be the case. Let me confirm with the project maintainer whether they actually want to enforce PascalCase on the subtree — if not, the current override is optimal and this issue can be closed. I'll circle back either way.

**Comment by TitikshaBansal** [comment]
Following up — I checked with @ltickett  (the project maintainer who raised the original discussion), and the answer to your question is yes: the project *does* want to enforce PascalCase across the subtree.

Concrete case:

- **Goal**: enforce `pascalCase` on filenames **and** directory names for all `.vue` files under `app/javascript/`.
- **Constraint**: the top-level `app/` folder is externally mandated by Rails / Vite Ruby conventions (Vite's entrypoint resolution, ActiveSupport autoloading, etc.) and cannot be renamed.
- **Plan**: the maintainer intends to rename the current lowercase subdirs (`common`, `pages`, `entrypoints`, `graphql`, `spec`, `stylesheets`) to PascalCase to comply with the standard.
- **Current workaround** (`checkDirectories: false`): gives up directory-name enforcement entirely. A future PR could reintroduce a lowercase subdir (say `app/javascript/utils/` because that's what a contributor is used to) with no lint error, and consistency degrades over time.
- **With `ignoreDirectories: [/^app$/]`**: the exact goal is expressible — `app` exempt as the one immovable segment, PascalCase enforced on every other directory in the subtree so drift can't sneak back in.

The workaround is fine as a short-term measure, but as a lint policy it's strictly weaker than what the project actually wants to enforce.

Happy to send a PR — the change is small (schema addition to both `anyOf` blocks, parse the option like `ignore`, skip matching segments in the `checkDirectories` loop, tests, docs). Let me know if Option A (`ignoreDirectories`) or Option B (change `ignore` semantics to skip only the directory check) is preferred.

**Comment by TitikshaBansal** [comment]
@sindresorhus is there any update or recommendation on this from your side?

**Comment by sindresorhus** [comment]
Thanks, that clarifies why `checkDirectories: false` is insufficient. Before deciding on the API, could you clarify two points?

1. The rule also checks the `javascript` segment. Will `app/javascript/` become `app/Javascript/`, or should `app/javascript` be treated as the root below which directory checking starts?
2. Should an exception apply to every directory named `app` at any depth, or only to specific root paths? For example, should a nested `app` directory also be exempt?

I think a root-based option may model this better than `ignoreDirectories`:

```js
directoryRoots: [
	'app/javascript',
	/^packages\/[^/]+\/src$/u,
]
```

Each entry would match an ancestor path relative to ESLint's working directory. The matched root and its ancestors would not be checked, while all descendant directories and filenames would be checked. If multiple roots match, the deepest one would win. If none match, the current behavior would remain.

For this project, `directoryRoots: ['app']` would exempt only `app` and still check `javascript`, while `directoryRoots: ['app/javascript']` would begin checking below the source root. This also supports multiple roots without globally exempting every directory named `app`.

Would those semantics match the intended policy?

**Comment by lonix1** [comment]
I'm using SvelteKit, rather than Vue/Rails, and have a very similar problem. I believe the same problem also exists in React/Next.js.

For SvelteKit, the conventions are:

- directories are camel cased: `src/foo/barBaz/quxQuux/`
- file-based routes are camel cased and prefixed by plus; there is finite set of such files, viz. `+page.svelte`, `+page.server.js|ts`, `+layout.svelte`, `+layout.server.js|ts`, `+error.svelte`, `+server.js|ts`
- components are pascal cased; e.g. `Profile.svelte`, `BillingHistory.svelte`

The route and component files can exist at any level of the directory hierarchy, so one cannot use two config blocks to target them separately.

This is the best possible option, but it's still invalid:

```js
{
  files: ['**/*.js', '**/*.ts'],
  // ...
  rules: {
    'unicorn/filename-case': ['error', { case: 'camelCase', checkDirectories: true }],
  }
},

{
  files: ['**/*.svelte', '**/*.svelte.js', '**/*.svelte.ts'],
  // ...
  rules: {
    // ...various svelte plugins and rules
    'unicorn/filename-case': [
      'error',
      {
        case: 'pascalCase',
        checkDirectories: false,                    // <---
        ignore: [                                   // <--- declare routes as exceptions, so components are the "norm"
          /^\+page(\.server)?\.(svelte|js|ts)$/,
          /^\+layout(\.server)?\.(svelte|js|ts)$/,
          /^\+error\.svelte$/,
          /^\+server\.(js|ts)$/,
        ],
      },
    ],
  }
},

{ // .svelte.js|ts hold "runes" logic, not components, so revert to camelCase
  files: ['**/*.svelte.js', '**/*.svelte.ts'],
  rules: {
    'unicorn/filename-case': ['error', { case: 'camelCase', checkDirectories: true }],
  },
},
```

Pros:

- The route files are listed as an exhaustive set of exceptions
- The component files thus become the norm using pascal case

Cons:

- The subdirectories must thus _also_ be pascal case, which breaks the convention; one could use `checkDirectories: false` as above, but that would prevent consistency in directories, and would allow anything like `src/foo/BAR/baz-qux/QuX`.

I don't know how to solve this problem, I just wanted to show that it's more common than one would think, across different environments.
